File: Serverless Functions/AWS/java17/src/main/java/com/benjaminbork/thesis/LambdaFunction.java
package com.benjaminbork.thesis;

import com.amazonaws.services.lambda.runtime.Context;
import com.amazonaws.services.lambda.runtime.RequestHandler;

import java.util.HashMap;
import java.util.Map;

import org.json.JSONArray;
import org.json.JSONObject;

public class LambdaFunction implements RequestHandler<Map<String, Object>, Map<String, Object>> {
    private static boolean isColdStart = true;

    @Override
    public Map<String, Object> handleRequest(Map<String, Object> input, Context context) {
        long startTime = System.currentTimeMillis();
        int cpuCores = Runtime.getRuntime().availableProcessors();
        Map<String, Object> responseMap = new HashMap<>();

        try {
            String body = input.get("body").toString();
            JSONObject bodyJson = new JSONObject(body);
            JSONObject payload = bodyJson.getJSONObject("payload");
            JSONArray matrixA = payload.getJSONArray("matrixA");
            JSONArray matrixB = payload.getJSONArray("matrixB");

            int matrixARows = matrixA.length();
            int matrixACols = matrixA.getJSONArray(0).length();
            int matrixBRows = matrixB.length();
            int matrixBCols = matrixB.getJSONArray(0).length();

            if (matrixACols != matrixBRows) {
                throw new Exception(
                        "Matrix A columns must be equal to Matrix B rows for multiplication");
            }
            int[][] result = new int[matrixARows][matrixBCols];

            // 2. Execute the matrix multiplication
            for (int i = 0; i < matrixARows; i++) {
                for (int j = 0; j < matrixBCols; j++) {
                    for (int k = 0; k < matrixACols; k++) {
                        result[i][j] += matrixA.getJSONArray(i).getInt(k) *
                                matrixB.getJSONArray(k).getInt(j);
                    }
                }
            }

            // 3. Collect all metrics
            boolean wasColdStart = isColdStart;
            LambdaFunction.isColdStart = false;
            long endTime = System.currentTimeMillis();
            long executionTime = endTime - startTime;

            responseMap.put("statusCode", 200);
            responseMap.put("message", "Matrix mulitplication executed successfully");
            responseMap.put("cloudProvider", "AWS Lambda");
            responseMap.put("allocatedMemory", 512);
            responseMap.put("cpuCores", cpuCores);
            responseMap.put("region", System.getenv("AWS_REGION"));
            responseMap.put("runTime", "Java 17");
            responseMap.put("functionName", "matrix-multiplication");
            responseMap.put("isColdStart", wasColdStart);
            responseMap.put("executionTime", executionTime);

        } catch (Exception e) {
            responseMap.put("statusCode", 400);
            responseMap.put("message", e.getMessage());
        }

        JSONObject responseJson = new JSONObject(responseMap);
        Map<String, Object> response = new HashMap<>();
        response.put("statusCode", responseJson.getInt("statusCode"));
        response.put("body", responseJson.toString());

        return response;

    }
}

File: Serverless Functions/AWS/python3.9/lambda_function.py
import os
import time
import json


is_cold_start = True

def lambda_handler(event, context):
    global is_cold_start
    start_time = time.time() * 1000
    cpu_cores = os.cpu_count() 

    try:
        # 1. Prepare the matix multiplication
        body = json.loads(event['body'])
        payload = body['payload']
        matrix_A = payload.get('matrixA', [])
        matrix_B = payload.get('matrixB', [])

        matrix_A_Rows = len(matrix_A)
        matrix_A_Columns = len(matrix_A[0])
        matrix_B_Rows = len(matrix_B)
        matrix_B_Columns = len(matrix_B[0])
        
        if (matrix_A_Columns != matrix_B_Rows):
            raise ValueError('Matrix A columns must be equal to Matrix B rows for multiplication')
    

        result = [[0 for _ in range(matrix_A_Columns)] for _ in range(matrix_B_Columns)] 


        # 2. Execute the matrix multiplication
        for i in range(matrix_A_Rows):
            for j in range(matrix_B_Columns):
                for k in range(matrix_B_Rows):
                    result[i][j] += matrix_A[i][k] * matrix_B[k][j]

        # 3. Collect all metrics
        was_cold_start = is_cold_start
        is_cold_start = False

        end_time = time.time() * 1000
        execution_time = end_time - start_time
        

        response = {
            'message': 'Matrix mulitplication executed successfully',
            'cloudProvider': 'AWS Lambda',
            'allocatedMemory': 512,
            'cpuCores': cpu_cores,
            'region': os.environ.get('AWS_REGION'),
            'runTime': 'python3.9',
            'functionName': 'matrix-multiplication',
            'isColdStart': was_cold_start,
            'executionTime': execution_time,
        }

        return {
            'statusCode': 200,
            'body': json.dumps(response)
        }

    except Exception as e:
        return {
            'statusCode': 400,
            'body': json.dumps({
                'message': 'Matrix mulitplication failed',
                'error': str(e),
            })
        }
